drafts/2019-03-15-integration-testing-with-testcontainers.adoc

```diff
@@ -0,0 +1,317 @@
+                .withNetwork(network)
+                .withNetworkAliases("testmongo");
+
+@ClassRule
+public static GenericContainer libertyContainer = new GenericContainer("my-service")
+                .withExposedPorts(9080)
+                .withLogConsumer(new Slf4jLogConsumer(LOGGER))
+                .waitingFor(Wait.forHttp("/myservice"))
+                .withNetwork(network)
+                .withEnv("MONGO_HOSTNAME", "testmongo")
+                .withEnv("MONGO_PORT", "27017");
+----
+
+Note that we can also set environment variables on a `GenericContainer` -- exactly as you would do in production with your docker containers. In this case, the application may be configured to locate the MongoDB server using the `MONGO_HOSTNAME` and `MONGO_PORT` environment variables.
+
+== Putting it all together
+
+Even though Testcontainers was originally intended for setting up external resources, I have found it to be extremely useful for testing my application's code in a docker container as well. Especially if you are already packaging your application into a docker container, writing integration tests with this approach is very simple, and can be done in pure Java.
+
+Once we have the container rules set up in our test class, we can write lots of simple JUnit tests that invoke HTTP requests on the docker container:
+[source,java]
+----
+@Test
+public void testGetPerson() {
+    Long bobId = personSvc.createPerson("Bob", 24);
+    Person bob = personSvc.getPerson(bobId);
+    assertEquals("Bob", bob.name);
+    assertEquals(24, bob.age);
+    assertNotNull(bob.id);
+}
+
+@Test
+public void testGetAllPeople() {
+    Long person1Id = personSvc.createPerson("Person1", 1);
+    Long person2Id = personSvc.createPerson("Person2", 2);
+
+    Person expected1 = new Person("Person1", 1, person1Id);
+    Person expected2 = new Person("Person2", 2, person2Id);
+
+    Collection<Person> allPeople = personSvc.getAllPeople();
+    assertTrue("Expected at least 2 people to be registered, but there were only: " + allPeople,
+               allPeople.size() >= 2);
+    assertTrue("Did not find person " + expected1 + " in all people: " + allPeople,
+               allPeople.contains(expected1));
+    assertTrue("Did not find person " + expected2 + " in all people: " + allPeople,
+               allPeople.contains(expected2));
+}
+
+// HTTP error code responses automatically get mapped to Exceptions with the type-safe rest client
+@Test(expected = NotFoundException.class)
+public void testGetUnknownPerson() {
+    personSvc.getPerson(-1L);
+}
+
+@Test(expected = BadRequestException.class)
+public void testCreateBadPersonNullName() {
+    personSvc.createPerson(null, 5);
+}
+----
+
+== Try it out!
+
+All of the code in this blog post is https://github.com/aguibert/basic-liberty/tree/testcontainers-prototype[available on Github]. Simply clone the repo and follow the steps in the README to run it locally.
```

Comment:
Excellent!  You might want to reference this github repo earlier in the article if people want to follow along with all of the code and not just the snippets.